### PR TITLE
feat: add domain padding

### LINKED ISSUE #706: Domain margins
**Is your feature request related to a problem? Please describe.**
Add margin to domain bounds

related to #510

### PR DESCRIPTION
**Pull request by nickofthyme**
## Summary

fixes #706

- add margin option to domain
- allow unconstrained padded domains, basically allow domains with padding to go beyond the zero baseline. By default, the domain is constrained to zero.
- create specific y domain type
- **deprecate** `yScaleToDataExtent` in favor of `domain.fit`

### Checklist

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios

### CODE REVIEW

File: api/charts.api.md [hunk 1]
```diff
@@ -1598,6 +1599,12 @@ export interface XYChartSeriesIdentifier extends SeriesIdentifier {
     yAccessor: string | number;
 }
 
+// Warning: (ae-forgotten-export) The symbol "YDomainBase" needs to be exported by the entry point index.d.ts
+// Warning: (ae-missing-release-tag) "YDomainRange" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply api/charts.api.md:1615]: I wasn't due to the others one being the same but you are right, I'll track it down and fix it.

[the remark this answers is not in the document]
Comment (nickofthyme) [reply api/charts.api.md]: e0d8264

[the remark this answers is not in the document]
Comment (markov00) [reply api/charts.api.md]: thanks nick, it wasn't necessary to fix every warning, just the one created with these changes :D

[the remark this answers is not in the document]
Comment (nickofthyme) [reply api/charts.api.md]: I used a regex that made it quick 🤷

File: src/utils/domain.ts [hunk 2]
```diff
@@ -43,65 +43,78 @@ export function computeOrdinalDataDomain(
...
   return [0, 0];
 }
 
 /** @internal */
 export function computeContinuousDataDomain(
   data: any[],
   accessor: (n: any) => number,
-  scaleToExtent = false,
-  fitToExtent = false,
+  domainOptions?: YDomainRange | null,
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/utils/domain.ts:116]: 74d3672

File: src/chart_types/xy_chart/domains/y_domain.ts [hunk 3]
```diff
@@ -84,35 +84,39 @@ function mergeYDomainForGroup(
...
+    // TODO remove when removing yScaleToDataExtent
+    const shouldScaleToExtent = groupSpecs.stacked.some(({ yScaleToDataExtent }) => yScaleToDataExtent)
+      || groupSpecs.nonStacked.some(({ yScaleToDataExtent }) => yScaleToDataExtent);
+    if (customDomain?.fit !== true && shouldScaleToExtent) {
+      if (!customDomain) {
+        customDomain = {};
+      }
+
+      customDomain.fit = true;
+    }
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/chart_types/xy_chart/domains/y_domain.ts]: 74d3672

File: src/utils/domain.ts [hunk 4]
```diff
@@ -43,65 +43,83 @@ export function computeOrdinalDataDomain(
...
+
+    if (/\d+%$/.test(padding.trim())) {
+      const paddingPercent = Number.parseInt(padding.trim().slice(0, -1), 10);
+      const delta = Math.abs(end - start);
+      computedPadding = delta * (paddingPercent / 100);
+    }
+    const num = Number.parseFloat(padding);
+
+    computedPadding = num && !isNaN(num) ? num : 0;
+  } else {
```
Comment (markov00) [inline src/utils/domain.ts:64]: At line 63 of domain.ts, you are overwriting the computedPadding the padding expressed in percentage.
If in percentage the computedPadding enters the if at line 56 but then the computed padding is overwritten by its float value parsed at line 61. This prevent the use of percentage strings as described in the docs

Comment (nickofthyme) [reply src/utils/domain.ts]: f088063

**Pull request merged**